src/overlay/OverlayManagerImpl.cpp

```diff
@@ -157,6 +157,58 @@ OverlayManagerImpl::PeersList::removePeer(Peer* peer)
     CLOG_WARNING(Overlay, "{}", REPORT_INTERNAL_BUG);
 }
 
+namespace
+{
+template <typename T>
+T
+getElement(typename std::vector<T>::iterator const& it)
+{
+    return *it;
+}
+
+template <typename T>
+T
+getElement(typename std::map<NodeID, T>::iterator const& it)
+{
+    return it->second;
+}
+
+template <typename C>
+void
+maybeDrop(C& peers, std::function<bool(Peer::pointer, uint32_t)> predicate,
+          uint32_t version, std::string const& reason)
+{
+    std::vector<Peer::pointer> peersToDrop;
+    for (auto it = peers.begin(); it != peers.end(); ++it)
+    {
+        auto peer = getElement<Peer::pointer>(it);
+        if (predicate(peer, version))
+        {
+            peersToDrop.emplace_back(peer);
+        }
+    }
+
+    for (auto& peer : peersToDrop)
+    {
+        // Drop will cleanup peer lists and remove peer references from overlay
+        // manager
+        peer->drop(reason, Peer::DropDirection::WE_DROPPED_REMOTE,
+                   Peer::DropMode::IGNORE_WRITE_QUEUE);
+    }
+}
+}
+
+void
+OverlayManagerImpl::PeersList::dropIf(
+    std::function<bool(Peer::pointer, uint32_t)> predicate, uint32_t version,
+    std::string const& reason)
+{
+    maybeDrop<decltype(OverlayManagerImpl::PeersList::mPending)>(
+        mPending, predicate, version, reason);
+    maybeDrop<decltype(OverlayManagerImpl::PeersList::mAuthenticated)>(
```

Comment:
you don't need to do all this templating stuff: you can just call `getPendingPeers` and `getRandomAuthenticatedPeers` (that also gives you a list), if you don't want to randomize for the later, you can refactor it.

Reply:
Updated.